src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,559 @@
+        assert_eq!(
+            reclocked_msgs,
+            &[(1, 1000), (2, 1000), (3, 2000), (4, 2000)]
+        );
+        assert!(batch[&PART_ID].is_empty());
+
+        // Advance forward but not enough to lose binding information
+        upper.insert(PART_ID, MzOffset::from(2));
+        assert_eq!(Antichain::from_elem(1000), operator.advance(&upper).await);
+
+        // Verify that we haven't lost information. Offset 2 should still be timestamped at 1000
+        batch.insert(
+            PART_ID,
+            vec![
+                (2, MzOffset::from(2)),
+                (3, MzOffset::from(3)),
+                (4, MzOffset::from(4)),
+            ],
+        );
+        upper.insert(PART_ID, MzOffset::from(5));
+        let reclocked_msgs = operator
+            .reclock(&mut batch, &upper)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(2, 1000), (3, 2000), (4, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Advance forward more, now we lose the ability to correctly reclock offsets < 3
+        upper.insert(PART_ID, MzOffset::from(3));
+        assert_eq!(Antichain::from_elem(2000), operator.advance(&upper).await);
+
+        // Attempt to do it anyway and verify that timestamps < 3 are assigned to timestamp 2000
+        batch.insert(PART_ID, vec![(2, MzOffset::from(2))]);
+        upper.insert(PART_ID, MzOffset::from(3));
+        let reclocked_msgs = operator
+            .reclock(&mut batch, &upper)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(2, 2000)]);
+        assert!(batch[&PART_ID].is_empty());
+    }
+
+    #[tokio::test(start_paused = true)]
+    #[ignore]
+    async fn test_concurrency() {
+        const PART_ID: PartitionId = PartitionId::None;
+
+        // Create two operators pointing to the same shard
+        let shared_shard = ShardId::new();
+        let mut op_a = create_test_operator(shared_shard).await;
+        let mut op_b = create_test_operator(shared_shard).await;
+
+        // Reclock a batch from one of the operators
+        let mut upper = HashMap::new();
+        let mut batch = HashMap::new();
+
+        // Reclock offsets 1 and 2 to timestamp 1000 from operator A
+        batch.insert(
+            PART_ID,
+            vec![(1, MzOffset::from(1)), (2, MzOffset::from(2))],
+        );
+        upper.insert(PART_ID, MzOffset::from(3));
+        let reclocked_msgs = op_a
+            .reclock(&mut batch, &upper)
+            .await
+            .flat_map(|(_, msgs)| msgs)
+            .collect_vec();
+        assert_eq!(reclocked_msgs, &[(1, 1000), (2, 1000)]);
+        assert!(batch[&PART_ID].is_empty());
+
+        // Advance the time by a lot
+        tokio::time::advance(Duration::from_secs(10)).await;
+
+        // Reclock the same offsets from operator B and verify we get the same result
```

Comment:
i would `advance` operator a to `3` before this, to show that UNTIL operator `advances`, its local state doesn't incorrectly reclock offset < 3`: that is, it maintains a correctly compacted snapshot that is durably stored in persist